diff --git a/util/testutil/context.go b/util/testutil/context.go
index 0c9e0f6..02a06e3 100644
--- a/util/testutil/context.go
+++ b/util/testutil/context.go
@@ -49,8 +49,8 @@ func (c *MockContext) Value(interface{}) interface{} {
 // MockContextErrAfter is a MockContext that will return an error after a certain
 // number of calls to Err().
 type MockContextErrAfter struct {
-	MockContext
 	count     atomic.Uint64
+	MockContext
 	FailAfter uint64
 }
 
